### PR TITLE
Add file descriptor boundary check

### PR DESCRIPTION
**Pull request by Yaxuan-w**
## Description

Fixes [#317](https://github.com/Lind-Project/lind_project/issues/317)

 

### Type of change

Found a bug when fixing libc errno returns: safeposix didn't have boundary check for file descriptors. Then, when users don't initialize fd, it will cause error since the value of the variable would depend on random data in memory.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- `make test` under /lind_project
- `cargo test --lib` under /lind_project/src/safeposix-rust

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -544,7 +545,8 @@ impl Cage {
     //------------------------------------FSTATFS SYSCALL------------------------------------
 
     pub fn fstatfs_syscall(&self, fd: i32, databuf: &mut FSData) -> i32 {
-        let unlocked_fd = self.filedescriptortable[fd as usize].read();
+        let tmp = self.get_filedescriptor(fd).unwrap();
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs:548]: i think we should have a better name than tmp for these, can you not call this plus the read() unlock on the same line?

Comment (Yaxuan-w) [reply src/safeposix/syscalls/fs_calls.rs:548]: `self.get_filedescriptor(fd).unwrap().read()` will cause lifetime error, so I separate them

Comment (rennergade) [reply src/safeposix/syscalls/fs_calls.rs:548]: ok lets change all the tmp's to 'checkedfd'

**Review by rennergade — APPROVED** [review]
This looks great. Fixes the bug and is easier to read.

**Review by RusherRG — APPROVED** [review]
The checks related to the range of the fd value make sense. I think we had seen a few segfaults because the checks weren't present, which should be fixed now. Nice fix :tada:

**Pull request merged**